File: mxrec-add-ons/tools/matmul_optimize/prof_matmul.py
# !/usr/bin/env python3
# -*- coding: utf-8 -*-
# Copyright 2024. Huawei Technologies Co.,Ltd. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================

import os
import sys
from enum import Enum

import logging
import glob
import sqlite3
import torch
import numpy as np
import pandas as pd

CONST_16 = 16
RECORD_COUNT = 10
DATA_RANGE = (-1.0, 1.0)
WORKSPACE = os.getcwd()
WORKSPACE_PROF_LIST_FILE = os.path.join(WORKSPACE, 'log', f'prof_list.csv')
WORKSPACE_RESULT_FILE = os.path.join(WORKSPACE, 'log', f'res.csv')

os.environ["WORKSPACE"] = WORKSPACE
os.environ["ASCEND_GLOBAL_LOG_LEVEL"] = "3"
os.environ["ASCEND_SLOG_PRINT_TO_STDOUT"] = "0"


class CubeFormat(Enum):
    ND = 0
    NZ = 1
    ZN = 2
    ZZ = 3
    NN = 4
    VECTOR = 5

    def __repr__(self) -> str:
        return self.__name__


class OpParam:

    CSV_HEADER_B = "B"
    CSV_HEADER_M = "M"
    CSV_HEADER_K = "K"
    CSV_HEADER_N = "N"
    CSV_HEADER_TRANSA = "Transpose A"
    CSV_HEADER_TRANSB = "Transpose B"
    CSV_HEADER_FORMATA = "Data Format A"
    CSV_HEADER_FORMATB = "Data Format B"
    
    def __init__(self, b=0, m=0, k=0, n=0,
                 trans_a=False, trans_b=False, en_bias=False,
                 en_scale=False, en_residual=False,
                 format_a=CubeFormat.ND, format_b=CubeFormat.ND,
                 format_c=CubeFormat.ND) -> None:
        self.b = b
        self.m = m
        self.k = k
        self.n = n
        self.trans_a = trans_a
        self.trans_b = trans_b
        self.en_bias = en_bias
        self.en_scale = en_scale
        self.en_residual = en_residual
        self.format_a = format_a
        self.format_b = format_b
        self.format_c = format_c
    
    def __str__(self) -> str:
        return f"Shape: ({self.b}, {self.m}, {self.k}, {self.n}) \n" + \
               f"Transpose: A {self.trans_a}, B {self.trans_b} \n" + \
               f"(De)Quant: Bias {self.en_bias}, Scale {self.en_scale}, Residual {self.en_residual} \n" + \
               f"CubeFormat: format_a {self.format_a}, format_b {self.format_b}, format_c {self.format_c}"


def read_op_param(testcase):
    op_param = OpParam()
    op_param.b = int(testcase[OpParam.CSV_HEADER_B])
    op_param.m = int(testcase[OpParam.CSV_HEADER_M])
    op_param.k = int(testcase[OpParam.CSV_HEADER_K])
    op_param.n = int(testcase[OpParam.CSV_HEADER_N])
    op_param.trans_a = testcase[OpParam.CSV_HEADER_TRANSA]
    op_param.trans_b = testcase[OpParam.CSV_HEADER_TRANSB]
    op_param.format_a = testcase[OpParam.CSV_HEADER_FORMATA]
    op_param.format_b = testcase[OpParam.CSV_HEADER_FORMATB]
    return op_param


def check_and_add_outputpath_header(output_path):
    if not os.path.exists(output_path):
        flags = os.O_WRONLY  
        modes = stat.S_IWUSR | stat.S_IRUSR 
        with os.fdopen(os.open(output_path, flags, modes), 'w') as f:
            f.write("B,M,K,N,"
                    "task_duration,aic_time(us),aic_cube_time(us),"
                    "aic_cube_ratio,"
                    "aic_mte1_ratio,aic_mte2_ratio,aic_mte3_ratio,"
                    "aic_fixpipe_ratio"
                    "\n")


def write_profop_data(res_path: str, op_param: OpParam) -> None:

    task_duration = 0
    op_basic_info_csv = os.path.join(res_path, "OpBasicInfo.csv")
    if not os.path.exists(op_basic_info_csv):
        logging.info("[debuf] check file: %s", op_basic_info_csv)
        logging.info("Can not find OpBasicInfo.csv in %s", res_path)
        raise Exception(f"Can not find OpBasicInfo.csv in {res_path}")
    else:
        data = pd.read_csv(op_basic_info_csv)
        for _, row in data.iterrows():
            task_duration = row['Task Duration(us)']
            break

    aic_time_us = 0
    aic_cube_time_us = 0
    aic_cube_ratio = 0
    aic_mte1_ratio = 0
    aic_mte2_ratio = 0
    aic_mte3_ratio = 0
    aic_fixpipe_ratio = 0
    pipe_util_csv = os.path.join(res_path, "PipeUtilization.csv")
    if not os.path.exists(pipe_util_csv):
        logging.info("Can not find PipeUtilization.csv in %s", res_path)
        raise Exception(f"Can not find PipeUtilization.csv in {res_path}")
    else:
        data = pd.read_csv(pipe_util_csv)
        for _, row in data.iterrows():
            aic_time_us = row['aic_time(us)']
            aic_cube_time_us = row['aic_cube_time(us)']
            aic_cube_ratio = row['aic_cube_ratio']
            aic_mte1_ratio = row['aic_mte1_ratio']
            aic_mte2_ratio = row['aic_mte2_ratio']
            aic_mte3_ratio = row['aic_mte3_ratio']
            aic_fixpipe_ratio = row['aic_fixpipe_ratio']
            break

    output_path = WORKSPACE_RESULT_FILE
    check_and_add_outputpath_header(output_path)

    flags = os.O_WRONLY  
    modes = stat.S_IWUSR | stat.S_IRUSR 
    with os.fdopen(os.open(output_path, flags, modes), 'w') as f:
        f.write(f"{op_param.b},{op_param.m},{op_param.k},{op_param.n},"
                f"{task_duration},{aic_time_us},{aic_cube_time_us},"
                f"{aic_cube_ratio},"
                f"{aic_mte1_ratio},{aic_mte2_ratio},{aic_mte3_ratio},"
                f"{aic_fixpipe_ratio}"
                "\n")


def performance_test(csv: str = "test_case.csv", device_id: int = 0):
    data = pd.read_csv(WORKSPACE + "/csv/" + csv, delimiter=",")
    for _, testcase in data.iterrows():
        op_param = read_op_param(testcase)
        cmd = f"./test_aclnn_msprofop {op_param.m} {op_param.k} {op_param.n} {op_param.trans_a} " \
              f" {op_param.trans_b} {op_param.format_a} {op_param.format_b} {device_id}"
        os.system(f"msprof op --launch-count={5} --launch-skip-before-match=1 --application=\"{cmd}\" "\
                  f" --output={WORKSPACE}/prof | tee {WORKSPACE}/log/output.log")
        with open(f"{WORKSPACE}/log/output.log", 'r') as f:
            lines = f.readlines()  # 读取所有行
            prof_res_path = lines[-3].split(" ")[-1].replace('\n', '')
            logging.info("lines[-3]", lines[-3])

        output_path = WORKSPACE_PROF_LIST_FILE
        if not os.path.exists(output_path):
            logging.info("Create file: %s", output_path)
            flags = os.O_WRONLY  
            modes = stat.S_IWUSR | stat.S_IRUSR 
            with os.fdopen(os.open(output_path, flags, modes), 'w') as f:
                f.write("B,M,K,N,prof_res_path\n")    

        flags = os.O_WRONLY  
        modes = stat.S_IWUSR | stat.S_IRUSR 
        with os.fdopen(os.open(output_path, flags, modes), 'w') as f:
            f.write(f"{op_param.b},{op_param.m},{op_param.k},{op_param.n},"
                    f"{prof_res_path}\n")


def clean_space() :
    if os.path.exists(WORKSPACE + "/prof"):
        os.system("rm -rf {}/prof/*".format(WORKSPACE))
    if os.path.exists(WORKSPACE + "/log"):
        os.system("rm -f {}/log/*".format(WORKSPACE))


def gen_test_data(csv: str = "test_case.csv", gen_algo_type: int = 0):
    # 定义列名
    columns = [
                OpParam.CSV_HEADER_B,
                OpParam.CSV_HEADER_M,
                OpParam.CSV_HEADER_K,
                OpParam.CSV_HEADER_N,
                OpParam.CSV_HEADER_TRANSA,
                OpParam.CSV_HEADER_TRANSB,
                OpParam.CSV_HEADER_FORMATA,
                OpParam.CSV_HEADER_FORMATB
    ]

    # 创建一个空的DataFrame，用于存储数据
    df = pd.DataFrame(columns=columns)

    if gen_algo_type == 0:
        data = []
        n = 300
        # 逐行添加数据
        for i in range(n):
            m = i * 32 + 32
            k = i * 32 + 32
            n = i * 32 + 32
            data.append({OpParam.CSV_HEADER_B: 1,
                         OpParam.CSV_HEADER_M: m,
                         OpParam.CSV_HEADER_K: k,
                         OpParam.CSV_HEADER_N: n,
                         OpParam.CSV_HEADER_TRANSA: 1,
                         OpParam.CSV_HEADER_TRANSB: 1,
                         OpParam.CSV_HEADER_FORMATA: 0,
                         OpParam.CSV_HEADER_FORMATB: 0})
            
        logging.info("Append data. size: %d", len(data))
        new_df = pd.DataFrame(data)
        df = pd.concat([df, new_df], ignore_index=True)  # 忽略索引进行合并

        # 将DataFrame写入CSV文件
        df.to_csv(WORKSPACE + "/csv/" + csv, sep=",", index=False)
    else:
        logging.info("Not support gen_algo_type: %s", gen_algo_type)
        raise Exception(f"Not support gen_algo_type: {gen_algo_type}")


def analyze_result():
    prof_list_file = WORKSPACE_PROF_LIST_FILE
    if not os.path.exists(prof_list_file):
        logging.info("prof file not exist. %s", prof_list_file)
        raise Exception(f"prof file not exist. {prof_list_file}")
    data = pd.read_csv(prof_list_file)
    for _, row in data.iterrows():
        op_param = OpParam()
        op_param.b = row[OpParam.CSV_HEADER_B]
        op_param.m = row[OpParam.CSV_HEADER_M]
        op_param.k = row[OpParam.CSV_HEADER_K]
        op_param.m = row[OpParam.CSV_HEADER_N]
        prof_res_path = row["prof_res_path"].strip()
        write_profop_data(prof_res_path, op_param)

if __name__ == "__main__":
    op = str(sys.argv[1])
    logging.info("WORKSPACE: %s, OP: %s", WORKSPACE, op)
    if op == "clean":
        clean_space()
    elif op == "gen_test_data":
        input_output_csv_file = sys.argv[2]
        input_gen_algo_type = int(sys.argv[3])
        gen_test_data(input_output_csv_file, input_gen_algo_type)
    elif op == "prof":
        os.makedirs(WORKSPACE + "/log", exist_ok=True)
        os.makedirs(WORKSPACE + "/prof", exist_ok=True)
        input_csv_file = sys.argv[2]
        input_device_id = int(sys.argv[3])
        performance_test(input_csv_file, input_device_id)
    elif op == "analyze_result":
        analyze_result()
    else:
        logging.info("Unknow op, OP: [%s]", op)

File: mxrec-add-ons/tools/matmul_optimize/prof_matmul.cpp
/*
Copyright (c) Huawei Technologies Co., Ltd. 2024. All rights reserved.
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
        limitations under the License.
==============================================================================*/

#include "acl/acl.h"
#include "aclnnop/aclnn_mm.h"

#include <string>

void CALL_RT(int x)
{
    if (auto ret = (x != 0)) {
        std::error << "[ERROR] Failed to exec acl api " << #x << ", result: " << ret << std::endl; \
        throw 1;
    }
}

void CreateAclTensorShape(std::vector<int64_t> &shape,
                          std::vector<int64_t> &strides,
                          int nRow, int nCol,
                          int transpose)
{
    shape.clear();
    shape.push_back(nRow);
    shape.push_back(nCol);
    
    strides.clear();
    if (transpose) {
        strides.push_back(1);
        strides.push_back(nRow);
    } else {
        strides.push_back(nCol);
        strides.push_back(1);
    }
}

void CreateAclTensor2d(const std::vector<int64_t> &shape,
                       const std::vector<int64_t> &strides,
                       __fp16 *deviceAddr, aclTensor **tensor, int format)
{
    *tensor = aclCreateTensor(shape.data(),
        shape.size(),
        ACL_FLOAT16,
        strides.data(),
        0,
        format ==0 ? aclFormat::ACL_FORMAT_ND: aclFormat::ACL_FORMAT_FRACTAL_NZ,
        shape.data(),
        shape.size(),
        deviceAddr);
}

int main(int argc, char **argv)
{
    int64_t verifyLevel = 0;
    int64_t deviceId = 3;
    int deviceIdIndex = 8;
    deviceId = std::atoi(argv[device_id_index]);
    CALL_RT(aclInit(nullptr));
    CALL_RT(aclrtSetDevice(deviceId));

    aclrtStream stream;
    CALL_RT(aclrtCreateStream(&stream));

    int transA;
    int transB;
    int formatA;
    int formatB;
    int64_t m;
    int64_t n;
    int64_t k;
    int64_t lda;
    int64_t ldb;
    int64_t ldc;

    // 打印当前行的数据
    int index = 1;
    m = std::atoi(argv[index++]);
    k = std::atoi(argv[index++]);
    n = std::atoi(argv[index++]);
    transA = std::atoi(argv[index++]);
    transB = std::atoi(argv[index++]);
    formatA = std::atoi(argv[index++]);
    formatB = std::atoi(argv[index++]);

    printf("M: %d, K: %d, N: %d, transA: %d, transB: %d, formatA: %d, formatB: %d\n",
        m,
        k,
        n,
        transA,
        transB,
        formatA,
        formatB);

    RunMatmul(m, n, k, transA, transB, formatA, formatB);
    
    CALL_RT(aclrtResetDevice(deviceId));
    CALL_RT(aclFinalize());

    return 0;
}

void RunMatmul(int64_t m,
               int64_t n,
               int64_t k,
               int transA,
               int transB,
               int formatA,
               int formatB,)
{
    int64_t A_size = sizeof(__fp16) * m * k;
    int cubeSize = 16;
    if (formatA == 1) {
        A_size = sizeof(__fp16) *
                ((m + cubeSize - 1) / cubeSize * cubeSize) *
                ((k + cubeSize - 1) / cubeSize * cubeSize);
    }
    int64_t B_size = sizeof(__fp16) * n * k;
    if (formatB == 1) {
        B_size = sizeof(__fp16) *
                ((n + cubeSize - 1) / cubeSize * cubeSize) *
                ((K + cubeSize - 1) / cubeSize * cubeSize);
    }
    int64_t C_size = sizeof(__fp16) * m * n;

    __fp16 *dA = nullptr;
    __fp16 *dB = nullptr;
    __fp16 *dC = nullptr;

    aclTensor *tensorA = nullptr;
    aclTensor *tensorB = nullptr;
    aclTensor *tensorC = nullptr;

    CALL_RT(aclrtMalloc((void **)(&dA), A_size, ACL_MEM_MALLOC_HUGE_FIRST));
    CALL_RT(aclrtMalloc((void **)(&dB), B_size, ACL_MEM_MALLOC_HUGE_FIRST));
    CALL_RT(aclrtMalloc((void **)(&dC), C_size, ACL_MEM_MALLOC_HUGE_FIRST));

    std::vector<int64_t> shape;
    std::vector<int64_t> strides;

    CreateAclTensorShape(shape, strides, m, k, transA);
    CreateAclTensor2d(shape, strides, dA, &tensorA, formatA);

    CreateAclTensorShape(shape, strides, k, n, transB);
    CreateAclTensor2d(shape, strides, dB, &tensorB, formatB);

    CreateAclTensorShape(shape, strides, m, n, 0);
    CreateAclTensor2d(shape, strides, dC, &tensorC, 0);

    int8_t cubeMathType = 0;
    uint64_t workspaceSize = 0;
    aclOpExecutor *executor;
    // 调用aclnnMm第一段接口
    CALL_RT(aclnnMmGetWorkspaceSize(tensorA, tensorB, tensorC, cubeMathType, &workspaceSize, &executor));

    // 根据第一段接口计算出的workspaceSize申请device内存
    void *workspaceAddr = nullptr;
    if (workspaceSize > 0) {
        CALL_RT(aclrtMalloc(&workspaceAddr, workspaceSize, ACL_MEM_MALLOC_HUGE_FIRST));
    }
    // 调用aclnnMm第二段接口
    CALL_RT(aclnnMm(workspaceAddr, workspaceSize, executor, stream));

    // 算子结束后需要同步，才能测数据
    CALL_RT(aclrtSynchronizeStream(stream));

    if (workspaceSize > 0) {
        aclrtFree(workspaceAddr);
    }

    aclDestroyTensor(tensorA);
    aclDestroyTensor(tensorB);
    aclDestroyTensor(tensorC);

    CALL_RT(aclrtFree(dA));
    CALL_RT(aclrtFree(dB));
    CALL_RT(aclrtFree(dC));
}
